cmd/otelcol/main_test.go

```diff
@@ -14,3 +14,111 @@
 // limitations under the License.
 
 package main
+
+import (
+	"context"
+	"net"
+	"os"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRunFromCmdLine(t *testing.T) {
+	tests := []struct {
+		name     string
+		panicMsg string
+		skipMsg  string
+		args     []string
+		timeout  time.Duration
+	}{
+		{
+			name:    "agent",
+			args:    []string{"otelcol", "--config=config/collector/agent_config.yaml"},
+			timeout: 15 * time.Second,
+		},
+		{
+			name:    "gateway",
+			args:    []string{"otelcol", "--config=config/collector/gateway_config.yaml"},
+			timeout: 15 * time.Second,
+		},
+		{
+			name:    "default_discovery",
+			args:    []string{"otelcol", "--discovery", "--config=config/collector/agent_config.yaml"},
+			timeout: 30 * time.Second,
+		},
+		// Running the discovery with --dry-run in CI is not desirable because of the use of os.Exit(0) to end the execution.
+		// That prevents the test from releasing resources like ports. The test needs to catch the panic to not fail the test,
+		// however, the resources won't be properly released for the remaining tests that may use the same resources.
+		// Skipping the test by default but keeping it around to deliberate runs on dev box.
+		{
+			name:     "dry-run_discovery",
+			args:     []string{"otelcol", "--discovery", "--dry-run", "--config=config/collector/agent_config.yaml"},
+			timeout:  30 * time.Second,
+			panicMsg: "unexpected call to os.Exit(0) during test", // os.Exit(0) in the normal execution is expected for '--dry-run'.
+			skipMsg:  "Skipping this test by default because --dry-run uses os.Exit(0) to end the execution",
+		},
+	}
+
+	// Set execution environment
+	requiredEnvVars := map[string]string{
+		"SPLUNK_ACCESS_TOKEN":     "access_token",
+		"SPLUNK_HEC_TOKEN":        "hec_token",
+		"SPLUNK_REALM":            "test_realm",
+		"SPLUNK_LISTEN_INTERFACE": "127.0.0.1",
+		"NO_WINDOWS_SERVICE":      "true", // Avoid using the Windows service manager
+	}
+	for key, value := range requiredEnvVars {
+		os.Setenv(key, value)
+		defer os.Unsetenv(key)
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.skipMsg != "" {
+				t.Skip(tt.skipMsg)
+			}
+
+			// GH darwin runners don't have docker installed, skip discovery tests on them
+			// given that the docker_observer is enabled by default.
+			if runtime.GOOS == "darwin" && (tt.name == "default_discovery" || tt.name == "dry-run_discovery") {
+				if os.Getenv("GITHUB_ACTIONS") == "true" {
+					t.Skip("skipping discovery tests on darwin runners since they don't have docker installed")
+				}
+			}
+
+			testCtx, cancel := context.WithTimeout(context.Background(), tt.timeout)
+			defer cancel()
+
+			otelcolCmdTestCtx = testCtx
+			defer func() {
+				otelcolCmdTestCtx = nil
+			}()
+
+			// Wait for the ConfigServer to be down after the test.
+			defer waitForPort(t, "55554")
```

Comment:
excellent idea to check we recovered ports after or not. One caution though, sometimes it take a bit of time, not sure if we should sleep  a bit on the helper function to give it some time to recover.

Reply:
`waitForPort` by default waits 60 seconds for the port to be available, see code below.

Reply:
That said, with #5201 the port is expected to be released whenever run finished (except for `--dry-run` as commented in the test). Let me update that to put my fix to the test 😁

Reply:
Actually, I shouldn't do that because other tests can take the port and the check will fail and this will become flaky.

Reply:
sounds good. Let me know when you're done.

Reply:
Keeping as it is.